pkg/storage/tsdb/tombstones.go

```diff
@@ -0,0 +1,360 @@
+		return errors.Wrap(err, "serialize tombstone")
+	}
+
+	fullTombstonePath := path.Join(TombstonePath, getTombstoneFileName(tombstone.RequestID, tombstone.State))
+
+	// Check if the tombstone already exists for the same state. Could be the case the same request was made
+	// and is already in the middle of deleting series. Creating a new tombstone would restart
+	// the progress
+	tombstoneExists, err := m.TombstoneExists(ctx, tombstone.RequestID, tombstone.State)
+	if err != nil {
+		level.Error(m.logger).Log("msg", "unable to check if the same tombstone already exists", "requestID", tombstone.RequestID, "err", err)
+	} else if tombstoneExists {
+		return ErrTombstoneAlreadyExists
+	}
+
+	return errors.Wrap(m.bkt.Upload(ctx, fullTombstonePath, bytes.NewReader(data)), "upload tombstone")
+}
+
+func (m *TombstoneManager) TombstoneExists(ctx context.Context, requestID string, state BlockDeleteRequestState) (bool, error) {
+	fullTombstonePath := path.Join(TombstonePath, getTombstoneFileName(requestID, state))
+	exists, err := m.bkt.Exists(ctx, fullTombstonePath)
+
+	if exists || err != nil {
+		return exists, err
+	}
+
+	return false, nil
+}
+
+func (m *TombstoneManager) GetTombstoneByIDForUser(ctx context.Context, requestID string) (*Tombstone, error) {
+	found := []*Tombstone{}
+
+	for _, state := range AllDeletionStates {
+		filename := getTombstoneFileName(requestID, state)
+		exists, err := m.TombstoneExists(ctx, requestID, state)
+		if err != nil {
+			return nil, err
+		}
+
+		if exists {
+			t, err := m.ReadTombstone(ctx, path.Join(TombstonePath, filename))
+			if err != nil {
+				return nil, err
+			}
+			found = append(found, t)
+		}
+	}
+
+	if len(found) == 0 {
+		return nil, nil
+	}
+
+	// If there are multiple tombstones with the same request id but different state, want to return only the latest one
+	// The older states will be cleaned up by the compactor (TODO future PR).
+	return found[len(found)-1], nil
+
+}
+
+func (m *TombstoneManager) GetAllTombstonesForUser(ctx context.Context) ([]*Tombstone, error) {
+	// add all the tombstones to a map and check for duplicates: if a key exists with the same request ID (but two different states).
+	discovered := make(map[string]BlockDeleteRequestState)
+	err := m.bkt.Iter(ctx, TombstonePath, func(s string) error {
+		requestID, state, err := GetStateAndRequestIDFromTombstonePath(filepath.Base(s))
+		if err != nil {
+			return err
+		}
+
+		if prevState, exists := discovered[requestID]; !exists {
+			discovered[requestID] = state
+		} else {
+			// If there is more than one tombstone for a given request, we only want the latest state.
+			// States can move only in a specific direction, so the later state will always win.
+			orderA, err := state.GetStateOrder()
+			if err != nil {
+				return err
+			}
+			orderB, err := prevState.GetStateOrder()
+			if err != nil {
+				return err
+			}
+
+			// If the newer state found comes later in the order, then we replace the state in the map.
```

Comment:
In this case, I think it might be better to name the variables `newerStateOrder` and `prevStateOrder`; this do imply that if we rename `state` to `newerState` the readability will improve too :)